crates/telio-proxy/src/proxy.rs

```diff
@@ -283,27 +313,38 @@ impl StateEgress {
     }
 
     async fn send_outbound_data(&mut self, pk: PublicKey, msg: DataMsg) {
-        // if let Some((pk, msg)) = self.input.recv().await {
-        if let (Some((socket, None)), Some(wg_addr)) = (self.sockets.get(&pk), self.wg_addr) {
-            match socket.send_to(msg.get_payload(), wg_addr).await {
-                Ok(_) => {
-                    if self.conn_state.is_err() {
-                        self.conn_state = Ok(());
-                        telio_log_info!("Outbound conn telio -> WG working!");
+        if let Err(err) = self.send_data(pk, msg.clone()).await {
+            let mut should_resend = self.handle_error(err, Some(pk)).await;
+            while should_resend {
+                match self.send_data(pk, msg.clone()).await {
+                    Ok(_) => {
+                        should_resend = false;
+                    }
+                    Err(err) => {
+                        should_resend = self.handle_error(err, Some(pk)).await;
                     }
                 }
-                Err(err) => self.handle_error(err, Some(pk)).await,
             }
+        }
+    }
+
+    async fn send_data(&mut self, pk: PublicKey, msg: DataMsg) -> Result<(), std::io::Error> {
+        if let (Some((socket, None)), Some(wg_addr)) = (self.sockets.get(&pk), self.wg_addr) {
+            socket.send_to(msg.get_payload(), wg_addr).await.map(|_| {
+                if self.conn_state.is_err() {
+                    self.conn_state = Ok(());
+                    telio_log_info!("Outbound conn telio -> WG working!");
+                }
+            })
         } else {
-            self.handle_error(
-                std::io::Error::new(ErrorKind::AddrNotAvailable, "WG Address not available"),
-                Some(pk),
-            )
-            .await;
+            Err(std::io::Error::new(
+                ErrorKind::AddrNotAvailable,
+                "WG Address not available",
+            ))
         }
     }
 
-    async fn handle_error(&mut self, err: std::io::Error, pk: Option<PublicKey>) {
+    async fn handle_error(&mut self, err: std::io::Error, pk: Option<PublicKey>) -> bool {
```

Comment:
Nitpick: bool as a return type is pretty ambitious. Would be nice to replace it with a custom enum or something like that